src/Stratis.Bitcoin/Consensus/Rules/IConsensusRules.cs

```diff
@@ -69,6 +66,27 @@ public interface IConsensusRules
         RuleContext CreateRuleContext(ValidationContext validationContext, ChainedHeader consensusTip);
     }
 
+    public interface IConsensusRule
+    {
+        /// <summary>The engine this rule belongs to.</summary>
+        IConsensusRules ConsensusRules { get; set; }
+
+        /// <summary>
+        /// Allow a rule to initialize itself.
+        /// The rule can verify that other rules are present using the <see cref="IConsensusRules.Rules"/>.
+        /// The rule can internally initialize its state.
+        /// </summary>
+        void Initialize();
+
+        /// <summary>
+        /// Execute the logic in the current rule.
+        /// If the validation of the rule fails a <see cref="ConsensusErrorException"/> will throw.
+        /// </summary>
+        /// <param name="context">The context that has all info that needs to be validated.</param>
+        /// <returns>The execution task.</returns>
+        Task RunAsync(RuleContext context);
```

Comment:
We don't use returns tag if the method returns plain task.

Reply:
Please done review the last push I might rule it back.

Reply:
No problem